samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -19,258 +19,212 @@
-        debug("Completed writing checkpoint=%s into %s topic for system %s." format(checkpoint, checkpointTopic, systemName) )
-        loop.done
-      },
+    // register and start a consumer for the checkpoint topic
+    if (systemConsumer == null) {
+      val oldestOffset = getOldestOffset(checkpointSsp)
+      info(s"Starting checkpoint SystemConsumer from oldest offset $oldestOffset")
+      systemConsumer.register(checkpointSsp, oldestOffset)
+      systemConsumer.start
+    }
 
-      (exception, loop) => {
-        warn("Failed to write checkpoint log partition entry %s: %s. Retrying." format(key, exception))
-        debug("Exception detail:", exception)
-      }
-    )
+    if (failOnCheckpointValidation) {
+      info(s"Validating checkpoint stream")
+      systemAdmin.validateStream(checkpointSpec)
+    }
   }
 
   /**
-   * Read the last checkpoint for specified TaskName
-   *
-   * @param taskName Specific Samza taskName for which to get the last checkpoint of.
-   **/
+    * @inheritdoc
+    */
+  override def register(taskName: TaskName) {
+    debug(s"Registering taskName: $taskName ")
+    systemProducer.register(taskName.getTaskName)
+    taskNames += taskName
+  }
+
+  /**
+    * @inheritdoc
+    */
   override def readLastCheckpoint(taskName: TaskName): Checkpoint = {
     if (!taskNames.contains(taskName)) {
-      throw new SamzaException(taskName + " not registered with this CheckpointManager")
+      throw new SamzaException(s"Task: $taskName is not registered with this CheckpointManager")
     }
 
-    info("Reading checkpoint for taskName " + taskName)
+    info(s"Reading checkpoint for taskName $taskName")
 
     if (taskNamesToOffsets == null) {
-      info("No TaskName to checkpoint mapping provided.  Reading for first time.")
-      taskNamesToOffsets = readCheckpointsFromLog()
+      info("Reading checkpoints for the first time")
+      taskNamesToOffsets = readCheckpoints()
     } else {
-      info("Already existing checkpoint mapping.  Merging new offsets")
-      taskNamesToOffsets ++= readCheckpointsFromLog()
+      info("Updating existing checkpoint mappings")
+      taskNamesToOffsets ++= readCheckpoints()
+      info("Updating existing checkpoint mappings" + taskNamesToOffsets)
     }
 
     val checkpoint = taskNamesToOffsets.get(taskName).getOrElse(null)
 
-    info("Got checkpoint state for taskName %s: %s" format(taskName, checkpoint))
-
+    info(s"Got checkpoint state for taskName $taskName $checkpoint")
     checkpoint
   }
 
   /**
-   * Read through entire log, discarding changelog mapping, and building map of TaskNames to Checkpoints
-   */
-  def readCheckpointsFromLog(): Map[TaskName, Checkpoint] = {
-    val checkpoints = mutable.Map[TaskName, Checkpoint]()
+    * @inheritdoc
+    */
+  override def writeCheckpoint(taskName: TaskName, checkpoint: Checkpoint) {
+    val key = new KafkaCheckpointLogKey(grouperFactory, taskName, KafkaCheckpointLogKey.CHECKPOINT_TYPE)
```

Comment:
This can throw a validation exception. Maybe catch, add more context and rethrow as SamzaException?